From 8acff4e19351ada59baa61f4bbc2bc4f19377410 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Pali=20Roh=C3=A1r?= <pali.rohar@gmail.com>
Date: Sat, 11 Apr 2026 01:00:01 +0200
Subject: [PATCH] crt: Fix return value of fegetexcept(), fedisableexcept() and
 feenableexcept() functions

Function __mingw_controlfp() returns bits for disabled exceptions,
but function fegetexcept() must return bits for enabled exceptions.
So negate bits for return value.

Function __mingw_controlfp() when modifying cw returns new state,
but functions fedisableexcept() and feenableexcept() must return old state.
So before modifying cw via __mingw_controlfp(), call fegetexcept() to
calculate return value of both fedisableexcept() and feenableexcept()
functions. Also checks for input argument and return -1 when is invalid.

Signed-off-by: LIU Hao <lh_mouse@126.com>
(cherry picked from commit f36077b0a5ab0be062d95103a4182700b67ac8e4)
Signed-off-by: LIU Hao <lh_mouse@126.com>
---
 mingw-w64-crt/misc/fedisableexcept.c | 5 ++++-
 mingw-w64-crt/misc/feenableexcept.c  | 5 ++++-
 mingw-w64-crt/misc/fegetexcept.c     | 2 +-
 3 files changed, 9 insertions(+), 3 deletions(-)

diff --git a/mingw-w64-crt/misc/fedisableexcept.c b/mingw-w64-crt/misc/fedisableexcept.c
index 29b905e3d..5c3328d02 100644
--- a/mingw-w64-crt/misc/fedisableexcept.c
+++ b/mingw-w64-crt/misc/fedisableexcept.c
@@ -4,5 +4,8 @@
 
 int __cdecl fedisableexcept(int excepts)
 {
-  return __mingw_controlfp(excepts & FE_ALL_EXCEPT, excepts & FE_ALL_EXCEPT) & FE_ALL_EXCEPT;
+  if (excepts & ~FE_ALL_EXCEPT) return -1;
+  int old_excepts = fegetexcept();
+  __mingw_controlfp(excepts, excepts);
+  return old_excepts;
 }
diff --git a/mingw-w64-crt/misc/feenableexcept.c b/mingw-w64-crt/misc/feenableexcept.c
index 8a40bd5fd..7bd6f757f 100644
--- a/mingw-w64-crt/misc/feenableexcept.c
+++ b/mingw-w64-crt/misc/feenableexcept.c
@@ -4,5 +4,8 @@
 
 int __cdecl feenableexcept(int excepts)
 {
-  return __mingw_controlfp(0, excepts & FE_ALL_EXCEPT) & FE_ALL_EXCEPT;
+  if (excepts & ~FE_ALL_EXCEPT) return -1;
+  int old_excepts = fegetexcept();
+  __mingw_controlfp(0, excepts);
+  return old_excepts;
 }
diff --git a/mingw-w64-crt/misc/fegetexcept.c b/mingw-w64-crt/misc/fegetexcept.c
index ced9f50b0..6bbf8d1bd 100644
--- a/mingw-w64-crt/misc/fegetexcept.c
+++ b/mingw-w64-crt/misc/fegetexcept.c
@@ -4,5 +4,5 @@
 
 int __cdecl fegetexcept(void)
 {
-  return __mingw_controlfp(0, 0) & FE_ALL_EXCEPT;
+  return ~__mingw_controlfp(0, 0) & FE_ALL_EXCEPT;
 }

